plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClientFactory.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.bigquery;
+
+import com.google.api.gax.rpc.HeaderProvider;
+import com.google.auth.Credentials;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.trino.spi.connector.ConnectorSession;
+
+import javax.inject.Inject;
+
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class BigQueryClientFactory
+{
+    private final IdentityCacheMapping identityCacheMapping;
+    private final BigQueryCredentialsSupplier credentialsSupplier;
+    private final BigQueryConfig bigQueryConfig;
+    private final HeaderProvider headerProvider;
+    private final Cache<IdentityCacheMapping.IdentityCacheKey, BigQueryClient> clientCache;
+
+    @Inject
+    public BigQueryClientFactory(IdentityCacheMapping identityCacheMapping, BigQueryCredentialsSupplier credentialsSupplier, BigQueryConfig bigQueryConfig, HeaderProvider headerProvider)
+    {
+        this.identityCacheMapping = requireNonNull(identityCacheMapping, "identityCacheMaping is null");
+        this.credentialsSupplier = requireNonNull(credentialsSupplier, "credentialsSupplier is null");
+        this.bigQueryConfig = requireNonNull(bigQueryConfig, "bigQueryConfig is null");
+        this.headerProvider = requireNonNull(headerProvider, "headerProvider is null");
+
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(bigQueryConfig.getServiceCacheTtl().toMillis(), MILLISECONDS);
```

Comment:
Is the reason for the cache to avoid client creation overhead?

Reply:
What happens if a query runs longer than the cache duration?

Reply:
Yes. It takes around a second to create it

Reply:
@hashhar we will create new instance. This will add some overhead but not significat as we pass instances to the `ReadSessionCreator` (and cache will just drop reference without closing client)